i6_models/parts/blstm.py

```diff
@@ -0,0 +1,78 @@
+from dataclasses import dataclass
+import torch
+from torch import nn
+
+from i6_models.config import ModelConfiguration
+
+
+@dataclass()
+class BlstmEncoderV1Config(ModelConfiguration):
+    """
+    Attributes:
+        num_layers: number of bi-directional LSTM layers, minimum 2
+        input_dim: input dimension size
+        hidden_dim: hidden dimension of one direction of LSTM, the total output size is twice of this
+        dropout: nn.LSTM supports internal Dropout applied between each layer of BLSTM (but not on input/output)
+        enforce_sorted: keep activated for ONNX-Export, requires that the lengths are sorted decreasing from longest
+
+            Sorting can for example be performed independent of the ONNX export in e.g. train_step:
+
+                audio_features_len, indices = torch.sort(audio_features_len, descending=True)
+                audio_features = audio_features[indices, :, :]
+                labels = labels[indices, :]
+                labels_len = labels_len[indices]
+    """
+
+    num_layers: int
+    input_dim: int
+    hidden_dim: int
+    dropout: float
+    enforce_sorted: bool = True
```

Comment:
Yes, this is a good default. Any model should be ONNX exportable per default. Also I still think it is better (perfomancewise) to sort the whole batch in the beginning, instead of doing it separately for the RNN stack and the loss computation.

And what does this have to do with RETURNN itself? This is purely on the side of the user if you want to do sorting by hand or not.

Reply:
As a user, I would expect that I could just use default settings and don't care about it. This would not be the case here.

Reply:
> Also, I don't think that ONNX exportability is such a high priority. ONNX will probably be replaced by sth else in the future anyway. And ONNX export is not needed for most cases except of RASR. And this sounds like it is really just a workaround for a current version of ONNX which might already be fixed next week or so. The other argument about performance is more relevant here.

This model is for doing experiments next week, not at some undefined future, at which we might decide to drop BLSTM in favor of other models or have a different implementations. I understood my task as: I should push a PyTorch BLSTM model that works with RASR in its current form, and was tested to give accurate WERs compared to the TF BLSTM within a RASR setup. And for this the only currently accepted way is via ONNX (making it the top priority).

As the users are from our chair (or affiliates to that), they expect that if they write a config with this model it can be exported for RASR.

Reply:
When you want to push sth here in this repo, I think it is relevant that it is in good shape, and not just useful for the experiments you do next week.

What happens when you have `enforce_sorted=True` but the seqs are actually not sorted?

Would it throw an error then? If that is the case, then this still maybe ok. But I would still argue that this is unexpected behavior for the common user, who likely would expect to be able to use sth like a `BlstmEncoder` just as-is.

Or would it perform some incorrect computation without throwing an error? If that is the case, then definitely `enforce_sorted=True` cannot be the default. That would be very dangerous, lead to hard-to-debug behavior.

If you don't know, this is sth important which should be checked first.

Or we just make the default as `enforce_sorted=False`. I think this is a more reasonable default. The average user can just use it as-is and don't care about details. And in case of ONNX export, the user would get an error and thus notice about the ONNX specific problem.

Or we would just not have a default if we cannot agree here.

I think I made my opinion clear here and probably cannot add much more to that. So the input of some other people here would be nice. @michelwi @christophmluscher @Marvin84 or so?

Reply:
> When you want to push sth here in this repo, I think it is relevant that it is in good shape, and not just useful for the experiments you do next week.

Our discussion is not about good shape, but about usage preference. I am just assuming that more than 50% of the cases somebody uses this encoder it will be used for RASR. 

> Would it throw an error then? 

Yes, it throws an error:

```
RuntimeError: `lengths` array must be sorted in decreasing order when `enforce_sorted` is True. You can pass `enforce_sorted=False` to pack_padded_sequence and/or pack_sequence to sidestep this requirement if you do not need ONNX exportability.
```

> Or we would just not have a default if we cannot agree here.

This is a valid option.

Reply:
I think this is something we should discuss today, since I think this trade-off will be a discussion in future .
On the other side: What is the use-case for this model? For now I know of Hybrid Baseline but do we actually use it for other experiments? Hybrid requires this export from my understanding so I would vote pro `True` default, but if there are other models which do not use it I would prefer `False` or nothing.

Reply:
> Or we would just not have a default if we cannot agree here.
Yes, I believe we decided that we want to only want to apply defaults very sparsely when there is a good reason and it's universally agreed upon. So if this one is controversial I would just not set a default value.

Reply:
Result of offline discussion: don't set default, but write in the docstring that the recommended value is True.